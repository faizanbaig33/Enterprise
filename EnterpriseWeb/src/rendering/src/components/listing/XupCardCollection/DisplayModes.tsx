import { ComponentRendering, Placeholder } from '@sitecore-jss/sitecore-jss-nextjs';
import classNames from 'classnames';
import React from 'react';
import { SliderWrapper } from 'src/helpers/SliderWrapper';

type FavoriteProducts = {
  favoriteProducts: string[];
};
type RenderingProps = {
  rendering: ComponentRendering;
};

type GridDisplayProps = {
  getLayoutClasses?: (numberOfCards: number, currentIndex: number) => string;
} & RenderingProps &
  FavoriteProducts;

type SliderDisplayProps = {
  sliderSettings: { [key: string]: any };
} & RenderingProps &
  FavoriteProducts;

export const GridDisplay = ({
  rendering,
  getLayoutClasses,
  favoriteProducts,
}: GridDisplayProps) => {
  return (
    <Placeholder
      name="cards"
      rendering={rendering}
      favoriteProducts={favoriteProducts}
      render={(cards) =>
        cards.map((card, index) => (
          <div
            key={index}
            className={classNames(
              'col-span-12',
              getLayoutClasses && getLayoutClasses(cards.length, index)
            )}
          >
            {card}
          </div>
        ))
      }
    />
  );
};

export const SliderDisplay = ({
  rendering,
  sliderSettings,
  favoriteProducts,
}: SliderDisplayProps) => {
  return (
    <div className="col-span-12">
      <Placeholder
        name="cards"
        rendering={rendering}
        favoriteProducts={favoriteProducts}
        render={(cards) => (
          <SliderWrapper sliderSettings={sliderSettings}>
            {cards.map((card, index) => (
              <div key={index} className={classNames('h-full md:px-xxs')}>
                {card}
              </div>
            ))}
          </SliderWrapper>
        )}
      />
    </div>
  );
};
