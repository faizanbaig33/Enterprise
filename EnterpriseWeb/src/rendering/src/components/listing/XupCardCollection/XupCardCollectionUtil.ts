export const ColumnSpans: { [cardsInRow: number]: string } = {
  1: 'md:col-span-12',
  2: 'md:col-span-6',
  3: 'md:col-span-4',
  4: 'md:col-span-3',
};

export const getLayoutClasses = (
  totalCards: number,
  currentIndex: number,
  maxCardsPerRow: number
): string => {
  let currentRow = 1; // currentRow will be used for grid display for dynamic column span calculation

  const totalRows = Math.ceil(totalCards / maxCardsPerRow);

  //If total cards are less than max cards per row
  if (totalCards <= maxCardsPerRow) return ColumnSpans[totalCards];
  //If all rows have equal cards
  else if (totalCards % maxCardsPerRow === 0) return ColumnSpans[maxCardsPerRow];
  else {
    // If row is filled with max cards than increase current row
    if (currentIndex > 0 && currentIndex % maxCardsPerRow === 0) currentRow++;

    // If current row is last row
    if (currentRow === totalRows) {
      const startCol = (12 / maxCardsPerRow) * (currentIndex % maxCardsPerRow) + 1;
      return `${ColumnSpans[maxCardsPerRow]} md:col-start-${startCol}`;
    } else {
      return ColumnSpans[maxCardsPerRow];
    }
  }
};
