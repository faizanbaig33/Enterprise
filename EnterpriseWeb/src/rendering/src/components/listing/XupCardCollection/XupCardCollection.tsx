// Components
import { withDatasourceCheck } from '@sitecore-jss/sitecore-jss-nextjs';
import classNames from 'classnames';
import { useTheme } from 'lib/context/ThemeContext';
import { getEnum } from 'lib/utils';
import { getBreakpoint, useCurrentScreenType } from 'lib/utils/get-screen-type';
import { ReactElement } from 'react';
import { Feature } from 'src/.generated/Feature.EnterpriseWeb.model';
import { Component } from 'src/helpers/Component';
import { GridDisplay, SliderDisplay } from './DisplayModes';
import { XupCardCollectionTheme } from './XupCardCollection.theme';
import { BodyCopy } from 'src/helpers/BodyCopy';
import { ButtonGroup } from 'src/helpers/ButtonGroup';
import { Headline } from 'src/helpers/Headline';
import { DefaultEditFrameButton, EditFrame } from 'src/helpers/EditFrame';

export type XupCardCollectionProps =
  Feature.EnterpriseWeb.Components.Listing.XupCardCollection.XupCardContainer;

type XupDisplayStyle = 'grid' | 'horizontal-scroll';

const ColumnSpans: { [cardsInRow: number]: string } = {
  1: 'md:col-span-12',
  2: 'md:col-span-6',
  3: 'md:col-span-4',
  4: 'md:col-span-3',
};

const XupCardCollection = (props: XupCardCollectionProps): JSX.Element => {
  const { currentScreenWidth } = useCurrentScreenType();
  const { themeData } = useTheme(XupCardCollectionTheme);

  const desktopDisplayStyle = getEnum<XupDisplayStyle>(props.fields.desktopDisplayStyle) || 'grid';
  const mobileDisplayStyle = getEnum<XupDisplayStyle>(props.fields.mobileDisplayStyle) || 'grid';

  const maxCardsPerRow: number = parseInt(getEnum<string>(props.fields.cardsPerRow) || '3');

  let currentRow = 1; // currentRow will be used for grid display for dynamic column span calculation

  const getLayoutClasses = (totalCards: number, currentIndex: number): string => {
    const totalRows = Math.ceil(totalCards / maxCardsPerRow);

    //If total cards are less than max cards per row
    if (totalCards <= maxCardsPerRow) return ColumnSpans[totalCards];
    //If all rows have equal cards
    else if (totalCards % maxCardsPerRow === 0) return ColumnSpans[maxCardsPerRow];
    else {
      // If row is filled with max cards than increase current row
      if (currentIndex > 0 && currentIndex % maxCardsPerRow === 0) currentRow++;

      // If current row is last row
      if (currentRow === totalRows) {
        const startCol = (12 / maxCardsPerRow) * (currentIndex % maxCardsPerRow) + 1;
        return `${ColumnSpans[maxCardsPerRow]} md:col-start-${startCol}`;
      } else {
        return ColumnSpans[maxCardsPerRow];
      }
    }
  };

  const sliderSettings = {
    dots: false,
    infinite: false,
    slidesToShow: maxCardsPerRow,
    responsive: [
      {
        breakpoint: getBreakpoint('md'),
        settings: {
          slidesToShow: 1,
          dots: true,
        },
      },
    ],
  };

  const renderCardColection = (): ReactElement => {
    if (currentScreenWidth < getBreakpoint('md')) {
      // Renderings for mobile devices
      if (mobileDisplayStyle === 'grid') {
        return (
          <GridDisplay rendering={props.rendering} favoriteProducts={props.favoriteProducts} />
        );
      } else {
        return (
          <SliderDisplay
            rendering={props.rendering}
            sliderSettings={sliderSettings}
            favoriteProducts={props.favoriteProducts}
          />
        );
      }
    } else {
      // Renderings for tablets and large screen devices
      if (desktopDisplayStyle === 'grid') {
        return (
          <GridDisplay
            rendering={props.rendering}
            getLayoutClasses={getLayoutClasses}
            favoriteProducts={props.favoriteProducts}
          />
        );
      } else {
        return (
          <SliderDisplay
            rendering={props.rendering}
            sliderSettings={sliderSettings}
            favoriteProducts={props.favoriteProducts}
          />
        );
      }
    }
  };

  return (
    <EditFrame
      title="Xup Card Collection Cards"
      dataSource={{ itemId: props.rendering.dataSource }}
      buttons={[{ ...DefaultEditFrameButton.edit, fields: ['ph-cards'] }]}
    >
      <Component variant="lg" dataComponent="listing/xupcardcollection" {...props}>
        <div className={classNames('col-span-10')}>
          <Headline classes={themeData.classes.headlineClass} {...props} />
          <BodyCopy classes={themeData.classes.bodyClass} {...props} />
          <ButtonGroup classes={themeData.classes.buttonGroupClass} {...props} />
        </div>
        {renderCardColection()}
      </Component>
    </EditFrame>
  );
};

export default withDatasourceCheck()<XupCardCollectionProps>(XupCardCollection);
