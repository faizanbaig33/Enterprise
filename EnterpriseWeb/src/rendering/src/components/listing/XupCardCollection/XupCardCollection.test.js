// Global
import { hasDataComponent, snapshot } from 'lib/jest/test-utils';
import 'lib/jest/match-media-mock';
// Local
import XupCardCollection from './XupCardCollection';
import defaultData from './XupCardCollection.mock-data';

it('renders correctly', () => {
  const component = snapshot(XupCardCollection, { componentProps: defaultData });
  hasDataComponent(component, 'listing/xupcardcollection');
});
