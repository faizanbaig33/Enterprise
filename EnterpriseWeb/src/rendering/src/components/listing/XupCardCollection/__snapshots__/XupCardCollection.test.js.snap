// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  data-component="listing/xupcardcollection"
  id="XupCardCollection"
>
  <div
    class="my-8 lg:mx-auto md:max-w-screen-lg theme-white"
  >
    <div
      class="section-grid grid grid-cols-2 md:grid-cols-12 px-m gap-s md:gap-s"
    >
      <div
        class="col-span-10"
      >
        <div
          class="items-top font-sans  text-sm-m md:text-m font-bold mb-xxs"
        >
          <h2>
            Xup Card Collection
          </h2>
        </div>
        <div
          class="text-body text-dark-gray font-regular mb-m"
        >
          <div
            class="body-copy"
          >
            <div>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc vulputate libero et velit interdum, ac aliquet odio mattis. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos. Curabitur tempus urna at turpis condimentum lobor.
            </div>
          </div>
        </div>
        <div
          class="flex flex-col md:flex-row items-start mb-s flex items-start md:flex-row md:items-center md:space-x-4"
        >
          <a
            class="flex w-fit items-center whitespace-nowrap font-sans text-text-link font-heavy text-theme-text hover:underline hover:decoration-secondary hover:underline-offset-8 disabled:border-gray disabled:text-gray mr-m"
            href="/"
            role="link"
            tabindex="0"
            title=""
          >
            Call to Action
            <span
              class="ml-xxs"
            />
          </a>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4"
      >
        <div
          style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
        >
          <h2>
            GenericCard
          </h2>
          <p>
            JSS component is missing React implementation. See the developer console for more information.
          </p>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4"
      >
        <div
          style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
        >
          <h2>
            GenericCard
          </h2>
          <p>
            JSS component is missing React implementation. See the developer console for more information.
          </p>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4"
      >
        <div
          style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
        >
          <h2>
            GenericCard
          </h2>
          <p>
            JSS component is missing React implementation. See the developer console for more information.
          </p>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4"
      >
        <div
          style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
        >
          <h2>
            GenericCard
          </h2>
          <p>
            JSS component is missing React implementation. See the developer console for more information.
          </p>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4"
      >
        <div
          style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
        >
          <h2>
            GenericCard
          </h2>
          <p>
            JSS component is missing React implementation. See the developer console for more information.
          </p>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4"
      >
        <div
          style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
        >
          <h2>
            GenericCard
          </h2>
          <p>
            JSS component is missing React implementation. See the developer console for more information.
          </p>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4 md:col-start-1"
      >
        <div
          style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
        >
          <h2>
            GenericCard
          </h2>
          <p>
            JSS component is missing React implementation. See the developer console for more information.
          </p>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4 md:col-start-5"
      >
        <div
          style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
        >
          <h2>
            GenericCard
          </h2>
          <p>
            JSS component is missing React implementation. See the developer console for more information.
          </p>
        </div>
      </div>
    </div>
  </div>
</section>
`;
